pkg/controllers/management/noderemove/nodes.go

```diff
@@ -0,0 +1,106 @@
+package noderemove
+
+import (
+	"context"
+
+	v3 "github.com/rancher/rancher/pkg/generated/norman/management.cattle.io/v3"
+	"github.com/rancher/rancher/pkg/ref"
+	"github.com/rancher/rancher/pkg/types/config"
+	"github.com/sirupsen/logrus"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+const (
+	PleaseKillMeAnnotation = "nodes.pkds.it/please-kill-me"
+)
+
+func Register(ctx context.Context, management *config.ManagementContext) {
+	nprc := &nodePoolRemoveController{
+		nodePoolController: management.Management.NodePools("").Controller(),
+		nodePools:          management.Management.NodePools(""),
+		nodeLister:         management.Management.Nodes("").Controller().Lister(),
+	}
+
+	management.Management.NodePools("").AddLifecycle(ctx, "nodepool-noderemove", nprc)
+}
+
+// NodePool Lifecycle
+
+type nodePoolRemoveController struct {
+	nodePoolController v3.NodePoolController
+	nodePools          v3.NodePoolInterface
+	nodeLister         v3.NodeLister
+}
+
+func (n *nodePoolRemoveController) Create(obj *v3.NodePool) (runtime.Object, error) {
+	return obj, nil
+}
+
+func (n *nodePoolRemoveController) Remove(obj *v3.NodePool) (runtime.Object, error) {
+	return obj, nil
+}
+
+func (n *nodePoolRemoveController) Updated(obj *v3.NodePool) (runtime.Object, error) {
+	nodes, err := n.listNodes(obj)
+	if err != nil {
+		return nil, err
+	}
+
+	nodesToRemove := make([]*v3.Node, 0)
+
+	for _, node := range nodes {
+		if HasRemovalAnnotation(node) {
+			nodesToRemove = append(nodesToRemove, node)
+		}
+	}
+
+	if len(nodesToRemove) == 0 {
+		return obj, nil
+	}
+
+	updated, err := n.nodePools.GetNamespaced(obj.Namespace, obj.Name, metav1.GetOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	if updated.Spec.Quantity != len(nodes) {
+		logrus.Infof("Another update is in progress for %s, will skip for now", updated.Spec.HostnamePrefix)
+		n.nodePoolController.Enqueue(updated.Namespace, updated.Namespace)
+
+		return updated, nil
+	}
+
+	logrus.Infof("Found %d nodes to remove from %s", len(nodesToRemove), obj.Spec.HostnamePrefix)
+
+	desiredQuantity := len(nodes) - len(nodesToRemove)
```

Comment:
I believe `nodesToRemove` could be an int here

Reply:
I'm keeping track of the nodes in a list because the deletion "behaviour" might change in the future (it has changed before) and we might need references to the nodes